### DISCUSSION TITLE
Using ALTMODE_RELATIVE with 2D Lines

### DISCUSSION THREAD

**Discussion by miles-tbe** [pelicanmapping/osgearth#1855] [Support Q&A] [answered]
Hi,
I have few questions regarding how GeoPoints using ALTMODE_RELATIVE in conjunction with a GeoTransform->setPosition() call works. 

I currently have this kind of structure for 2D Lines being drawn:
```cpp
GeoTransform GeoLine = new GeoTransform;
osg::Geometry Line = new osg::Geometry();
osg::Group LineGroup = new osg::Group();
Line->setVertexArray(osg::Vec3Array());
GeoLine->addChild (Line);
GeoLine->setPosition (GeoPoint (GeoGraphicSRS, Lon, Lat, Alt, ALTMODE_RELATIVE));
LineGroup->addChild (GeoLine);
```

I have a lot of these osg::Geometry objects as children to the GeoTransform object and I am updating them constantly using an UpdateOperation on the viewer. The way it is set up is the GeoTransform is removed from the group at the beginning of the update function and then the function that draws the lines is called and the GeoTransform is re-added to the group which is apart of the MapNode elsewhere. I have a condition in this update function where it stops updating the data for these lines. When I do this, I do not see the the lines drawing with the relative altitude until after the data updating stops. So what I want to know is:

1) How is the ALTMODE_RELATIVE applied to children of GeoTransforms using a GeoPoint for its position? Would this account for the children to be drawn at their original altitude _first_ and then have to be adjusted the the relative altitude?

2) I have another update track where it only updates when receiving a message saying to do so. In this case I get the child lines drawing initially at their unadjusted altitudes and then flashing to the adjusted one while waiting for the next update message and when run at a high speed, it produces a flickering visual. This seems  consistent with my above findings and again I want to know if this is the expected outcome or how is it supposed to happen?

3) Lastly, I want to know if there is a way to co-op this using of GeoPoints with LineDrawables directly? The desired use case would be to draw LineDrawables with Lat, Lon, Altitude rather than ECEF xyz coordinates where the the altitude was also relative to the terrain elevation height.

Thanks,
Miles

**Comment by gwaldron** [comment, marked as answer]
Hello Miles.
The GeoTransform processes an ALTMODE_RELATIVE by sampling the terrain. In order to do that it just find a pointer to the terrain engine, which happens automatically during the next UPDATE traversal. If you don't want to wait, you can call GeoTransform::setTerrain to do it manually; then calling setPosition will have immediate effect. Otherwise there may be a one-frame delay. Maybe that's what you are seeing.

I must admit that I fail to grok your explanation of removing nodes, drawing them, re-inserting, and stopping updates. So I cannot comment on whether this setup is exacerbating the problem.

**Reply by miles-tbe** [reply]
Using setTerrain fixed my issue I was having with where the lines were being drawn. Thank you. Before I mark this as answered I just wanted to know if you had any information regarding my third (3) question above at all or should I just make that a separate post?
